etc/dovetail-edge-deploy.yml

```diff
@@ -0,0 +1,866 @@
+  #           Statement:
+  #             - Effect: Allow
+  #               Action:
+  #                 - "cloudwatch:DeleteAlarms"
+  #                 - "cloudwatch:DescribeAlarms"
+  #                 - "cloudwatch:DisableAlarmActions"
+  #                 - "cloudwatch:EnableAlarmActions"
+  #                 - "cloudwatch:PutMetricAlarm"
+  #               Resource: "*"
+
+  # Pipeline functions
+  PublishEdgeLambdaVersionsFunctionIamRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+            Action:
+              - sts:AssumeRole
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+      Path: "/"
+      Policies:
+        - PolicyName: LambdaVersionPolicy
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+              - Effect: Allow
+                Action:
+                  - lambda:PublishVersion
+                Resource:
+                  - !Ref StagingStitchLambdaFunctionArn
+                  - !Ref StagingBytesLambdaFunctionArn
+                  - !Ref ProductionStitchLambdaFunctionArn
+                  - !Ref ProductionBytesLambdaFunctionArn
+        - PolicyName: CodePipelinePolicy
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+              - Effect: Allow
+                Action:
+                  - codepipeline:PutJobSuccessResult
+                  - codepipeline:PutJobFailureResult
+                  - codepipeline:GetPipelineState
+                Resource: "*"
+  PublishEdgeLambdaVersionsFunction:
+    Type: AWS::Lambda::Function
+    Properties:
+      Code:
+        ZipFile: |
+          import boto3
+          import traceback
+          import os
+          from botocore.client import Config
+
+          code_pipeline = boto3.client('codepipeline')
+          lamb = boto3.client('lambda')
+
+
+          def put_job_success(job, message):
+              print('Putting job success')
+              print(message)
+              code_pipeline.put_job_success_result(jobId=job['id'])
+
+
+          def put_job_failure(job, message):
+              print('Putting job failure')
+              print(message)
+              code_pipeline.put_job_failure_result(
+                  jobId=job['id'],
+                  failureDetails={'message': message, 'type': 'JobFailed'})
+
+
+          # Tries to publish new versions of each Lambda function, from their $LATEST
+          # code and configuration. New versions are only created if a any changes have
+          # been made since the most recent version, so it is safe to call this
+          # multiple times
+          def publish_versions():
+              lamb.publish_version(FunctionName=os.environ['STAGING_BYTES_LAMBDA_FUNCTION_ARN'])
```

Comment:
Do these throw an error, if the code hasn't changed?

Reply:
No, it's a no op